### PR TITLE
Feature/468 add schema.org website markup

### LINKED ISSUE #468: schema.org/WebSite markup
WordLift handles the schema.org/Article markup for single post/pages. It might be useful also to provide schema.org/WebSite markup for the Homepage:

http://schema.org/WebSite

An example snippet:
```
{
  "@context": "http://schema.org",
  "@type": "WebSite",
  "name": "Your WebSite Name",
  "alternateName": "An alternative name for your WebSite",
  "url": "http://www.your-site.com",
  "publisher": ...
}
```

The _publisher_ property should be filled as we do for Articles.

Should we allow to disable the JSON-LD snippet though in cases where the _WebSite_ markup is already provided by another plugin (or by the theme).

References:
* https://developers.google.com/search/docs/data-types/sitename

We should consider adding the _Search Action_ at a later stage:
* https://developers.google.com/search/docs/data-types/sitelinks-searchbox

### PR DESCRIPTION
**Pull request by Stoyan0v**
Fixes #468 

#### Changes proposed in this Pull Request:
 * Change schema.org to [WebSite](http://schema.org/WebSite) for home and blog pages

Currently home page is uses schema.org/Article, while there is a schema.org/WebSite that can be used. In this PR there are two conditions that check if the home or blog page is being displayed and change the schema to WebSIte. It will also add default [Search Action].(https://developers.google.com/search/docs/data-types/sitelinks-searchbox)

There are also two filters (both documented in the code):
* `wordlift_disable_website_json_ld ` that allow you to disable the WebSite markup in case other plugin provides it.
* `wordlift_json_ld_search_url ` that allow you to change the default search url

#### Testing instructions for homepage:

* Activate the plugin and add your API key
* Create sample page
* Go to Settings > Reading and make that page homepage (Front page)
* Open the page and make sure that `application/ld+json` script has been changed to WebSite

#### Testing instructions for blogpage:

* Activate the plugin and add your API key
* Create sample page
* Go to Settings > Reading and make that page blog page (Posts page)
* Open the page and make sure that `application/ld+json` script has been changed to WebSite

Currently there are some errors related to publisher, that will be better to be addressed in another PR, so we can keep this one small.

website-schema-check

There are also three unit tests added that covered new schema and the two filters.

PS: Feedback welcome

### CODE REVIEW

**Comment by ziodave** [comment]
@Stoyan0v the PR looks great from every perspective 👏.

There are a couple of issues though:

- [ ] `[null]` JSON-LD 

I am testing it on a WP 4.2 test web site though and I get the following on the home page:

`<script type="application/ld+json">[null]</script>`

Which is the return value of the AJAX call to `/wp-admin/admin-ajax.php` with the following parameters:
* action:wl_jsonld
* id:0

My test web site is set to display "Your latest posts", which we need to support too:
image

- [ ] _publisher_ type

In your screenshot I see _Event_ as publisher type:
image

Can you try changing it to _Organization_ or _Person_ (these are the only 2 types supported for _[publisher](http://schema.org/publisher)_) and retest it with GSTT? This should solve the issues raised by GSTT.

**Comment by Stoyan0v** [comment]
Thank you for the feedback @ziodave.
After checking the code I realized that the logic rely on the fact that there always will be post id and if it's missing nothing is displayed. At first I decided to change the `Wordlift_Jsonld_Service` and `Wordlift_Post_To_Jsonld_Converter` logic to allow creating schema without id, but this requires a lot of modifications, that can break backward compatibility, so I decided to use new converter.

Here is what was changed:

#### Changes proposed in this Pull Request:
 * A new converter that extends `Wordlift_Post_To_Jsonld_Converter` has been created.
 * I've added a [flag](https://github.com/insideout10/wordlift-plugin/pull/578/files#diff-9bdc689754858a84ba71460ae2a798daR111) to settings, so that we will know when it's a home/blog page
* Then in [Wordlift_Jsonld_Service](https://github.com/insideout10/wordlift-plugin/pull/578/files#diff-5697434245d28a1cf04fa571a129e1b2R80) I've check for that flag and return the homepage info, instead empty json if there is no id.
* In `Wordlift_Website_Jsonld_Converter` we fallback to `Wordlift_Post_To_Jsonld_Converter` if there is an id and the we change the schema, so it will be accurate for homepage.
* If there is no ID we create a new schema from scratch.
* Then we apply the search action.
* There is a new test that makes sure there is correct schema for homepage even when is displays the "Latest posts"

The only thing missing when there is no id the the publisher.
Also the publisher type now works fine for me and it's "Person".

**Comment by ziodave** [comment]
Hello @Stoyan0v,

I think we're almost there, the code looks always really good, with formatting and comments just the way WordPress likes it.

I found some issues (the list might look long, but very small fixes are required I suppose):

### _publisher_ missing

Tested with: "Front page displays: Your latest posts"

I don't see the [publisher](http://schema.org/publisher) property anymore in the JSON-LD:
image

I think the issue is here:

```
	public function apply_website_info_to_json_ld( &$params ) {
		// Change values so they will be accurate for homepage

		// Change schema properties.
		$params['@type']       = 'WebSite';
		$params['headline']    = get_bloginfo( 'name' );
		$params['description'] = get_bloginfo( 'description' );
		$params['url']         = home_url( '/' );

	}
```

Probably adding a call to ```$this->set_publisher( $jsonld );``` will fix it.

The _publisher_ information is really what makes WordLift shine in this context, as it's the only one able to provide a real entity (not just a label), so it's important that we provide this property.

Do you have a unit test for this?

### JSON-LD array not required

When browsing a post/page, WL prints out many entities related to the post and to each other, therefore the JSON-LD is output as an array.

For the WebSite case I don't think we need an array since the JSON-LD response contains only the WebSite.

This code:
```
				// Set a reference to the website_converter.
				$website_converter = $this->website_converter;

				$jsonld = array( $website_converter->create_schema( $_REQUEST ) );

				// Send JSON-LD.
				wp_send_json( $jsonld );
```

Can probably be simplified to ```wp_send_json( $this->website_converter->create_schema( $_REQUEST ) );```

### WebSite shouldn't inherit the Article properties

Tested with: "Front page displays: A static page"

In this case I see that the JSON-LD inherits the Article properties and then changes the type:

image

Which happens here:
```
		// Check if we have home/blog page id.
		if ( isset( $request['id'] ) ) {
			$post_id = $request['id'];

			// Get the base JSON-LD and the list of entities referenced by this entity.
			$jsonld = parent::convert( $post_id, $references );
		} else {
			// Create new jsonld, because we don't have id.
			$jsonld = array(
				'@context' => 'http://schema.org',
			);
		}
```

This is actually not required and might even be counterproductive, since the [schema.org/Article](http://schema.org/Article) has many properties that do not apply to [schema.org/WebSite](http://schema.org/WebSite) or it's wrong to assume that they're pertain the WebSite, e.g.:
 * dateModified/datePublished/author/mainEntityOfPage/headline/description pertain the page, not the Web Site,
 * wordCount is not a property a Web Site

I think this part can be omitted then:
```
		// Check if we have home/blog page id.
		if ( isset( $request['id'] ) ) {
			$post_id = $request['id'];

			// Get the base JSON-LD and the list of entities referenced by this entity.
			$jsonld = parent::convert( $post_id, $references );
		} else {
```

### WebSite properties missing

The following properties are missing from the JSON-LD (see [here](https://developers.google.com/search/docs/data-types/sitename) for more details):

- [ ] _name_ with the Web Site name
- [ ] _alternativeName_ with the Web Site slogan

It's probably enough to fix this:
```
	public function apply_website_info_to_json_ld( &$params ) {
		// Change values so they will be accurate for homepage

		// Change schema properties.
		$params['@type']       = 'WebSite';
		$params['headline']    = get_bloginfo( 'name' );
		$params['description'] = get_bloginfo( 'description' );
		$params['url']         = home_url( '/' );

	}
```

Into this:
```
	public function apply_website_info_to_json_ld( &$params ) {
		// Change values so they will be accurate for homepage

		// Change schema properties.
		$params['@type']       = 'WebSite';
		$params['name']    = get_bloginfo( 'name' );
		$params['alternateName'] = get_bloginfo( 'description' );
		$params['url']         = home_url( '/' );

	}
```

**Comment by Stoyan0v** [comment]
Hey @ziodave, thanks again for the review. I've addressed the issues. Here are short comments to the point you have listed above:

>> publisher missing

Publisher will be now always added to Website schema. I've added nessasery unit test too.

>>JSON-LD array not required

Make sense. I've removed that part

>>WebSite shouldn't inherit the Article properties

Agree, even when there is an ID the website converted will create new schema, that doesn't inherit the Article. 

>>WebSite properties missing

WebSite schema now using the correct property names.

**Comment by ziodave** [comment]
Looks great, merged 👌
